File: src/main/java/com/fengsheng/phase/OnUseCard.java
package com.fengsheng.phase;

import com.fengsheng.Fsm;
import com.fengsheng.Player;
import com.fengsheng.ResolveResult;
import com.fengsheng.card.Card;
import com.fengsheng.protos.Common;

/**
 * 使用卡牌时，成为卡牌目标时
 */
public class OnUseCard implements Fsm {
    /**
     * 谁的回合
     */
    public Player whoseTurn;
    /**
     * 目标角色
     */
    public Player targetPlayer;
    /**
     * 出牌的人
     */
    public Player player;
    /**
     * 出的牌，有可能没出牌
     */
    public Card card;
    /**
     * 出的牌的类型
     */
    public Common.card_type cardType;
    /**
     * 遍历到了谁的技能
     */
    public Player askWhom;
    /**
     * 卡牌效果的结算函数
     */
    public Fsm resolveFunc;

    public OnUseCard(Player whoseTurn, Player player, Player targetPlayer, Card card, Common.card_type cardType, Player askWhom, Fsm resolveFunc) {
        this.whoseTurn = whoseTurn;
        this.player = player;
        this.targetPlayer = targetPlayer;
        this.card = card;
        this.cardType = cardType;
        this.askWhom = askWhom;
        this.resolveFunc = resolveFunc;
    }

    @Override
    public ResolveResult resolve() {
        ResolveResult result = whoseTurn.getGame().dealListeningSkill();
        return result != null ? result : new ResolveResult(new OnUseCardNext(this), true);
    }

    @Override
    public String toString() {
        return player + "使用" + card + "时";
    }

    private record OnUseCardNext(OnUseCard onUseCard) implements Fsm {
        @Override
        public ResolveResult resolve() {
            int askWhom = onUseCard.askWhom.location();
            Player[] players = onUseCard.askWhom.getGame().getPlayers();
            while (true) {
                askWhom = (askWhom + 1) % players.length;
                if (askWhom == onUseCard.player.location()) {
                    return onUseCard.resolveFunc.resolve();
                }
                if (players[askWhom].isAlive()) {
                    onUseCard.askWhom = players[askWhom];
                    return new ResolveResult(onUseCard, true);
                }
            }
        }
    }
}

File: src/main/java/phase/OnUseCard.kt
package com.fengsheng.phase

import com.fengsheng.*
import com.fengsheng.card.*
import com.fengsheng.protos.Common.cardimport

com.fengsheng.protos.Common.card_type
/**
 * 使用卡牌时，成为卡牌目标时
 */
class OnUseCard(
    /**
     * 谁的回合
     */
    var whoseTurn: Player,
    /**
     * 出牌的人
     */
    var player: Player,
    /**
     * 目标角色
     */
    var targetPlayer: Player?,
    /**
     * 出的牌，有可能没出牌
     */
    var card: Card?,
    /**
     * 出的牌的类型
     */
    var cardType: card_type,
    /**
     * 遍历到了谁的技能
     */
    var askWhom: Player,
    /**
     * 卡牌效果的结算函数
     */
    var resolveFunc: Fsm
) : Fsm {
    override fun resolve(): ResolveResult? {
        val result = whoseTurn.game.dealListeningSkill()
        return result ?: ResolveResult(OnUseCardNext(this), true)
    }

    override fun toString(): String {
        return player.toString() + "使用" + card + "时"
    }

    private class OnUseCardNext(onUseCard: OnUseCard) : Fsm {
        override fun resolve(): ResolveResult? {
            var askWhom = onUseCard.askWhom.location()
            val players = onUseCard.askWhom.game.players
            while (true) {
                askWhom = (askWhom + 1) % players.size
                if (askWhom == onUseCard.player.location()) {
                    return onUseCard.resolveFunc.resolve()
                }
                if (players[askWhom].isAlive) {
                    onUseCard.askWhom = players[askWhom]
                    return ResolveResult(onUseCard, true)
                }
            }
        }

        val onUseCard: OnUseCard

        init {
            this.card = card
            this.sendPhase = sendPhase
            this.r = r
            this.target = target
            this.card = card
            this.wantType = wantType
            this.r = r
            this.target = target
            this.card = card
            this.player = player
            this.card = card
            this.card = card
            this.drawCards = drawCards
            this.players = players
            this.mainPhaseIdle = mainPhaseIdle
            this.dieSkill = dieSkill
            this.player = player
            this.player = player
            this.onUseCard = onUseCard
        }
    }
}